src/TelegramClient.Core/Network/Tcp/TcpTransport.cs

```diff
@@ -33,67 +34,85 @@ public void Dispose()
-                Buffer.BlockCopy(bodyByte, 0, body, readBytes, availableBytes);
-                readBytes += availableBytes;
-            }
-            while (readBytes != packetLength - 12);
+                do
+                {
+                    var bodyByte = new byte[packetLength - 12];
+                    var availableBytes = await stream.ReadAsync(bodyByte, 0, neededToRead).ConfigureAwait(false);
 
-            var crcBytes = new byte[4];
-            var readCrcBytes = await stream.ReadAsync(crcBytes, 0, 4).ConfigureAwait(false);
-            if (readCrcBytes != 4)
-            {
-                throw new InvalidOperationException("Couldn't read the crc");
-            }
+                    neededToRead -= availableBytes;
+                    Buffer.BlockCopy(bodyByte, 0, body, readBytes, availableBytes);
+                    readBytes += availableBytes;
+                }
+                while (readBytes < packetLength - 12 && stream.CanRead && stream.DataAvailable);
+
+                var crcBytes = new byte[4];
+                var readCrcBytes = await stream.ReadAsync(crcBytes, 0, 4).ConfigureAwait(false);
+                if (readCrcBytes != 4)
+                {
+                    throw new InvalidOperationException("Couldn't read the crc");
+                }
 
-            var checksum = BitConverter.ToInt32(crcBytes, 0);
+                var checksum = BitConverter.ToInt32(crcBytes, 0);
 
-            var rv = new byte[packetLengthBytes.Length + seqBytes.Length + body.Length];
+                var rv = new byte[packetLengthBytes.Length + seqBytes.Length + body.Length];
 
-            Buffer.BlockCopy(packetLengthBytes, 0, rv, 0, packetLengthBytes.Length);
-            Buffer.BlockCopy(seqBytes, 0, rv, packetLengthBytes.Length, seqBytes.Length);
-            Buffer.BlockCopy(body, 0, rv, packetLengthBytes.Length + seqBytes.Length, body.Length);
-            var crc32 = new Crc32();
-            crc32.SlurpBlock(rv, 0, rv.Length);
-            var validChecksum = crc32.Crc32Result;
+                Buffer.BlockCopy(packetLengthBytes, 0, rv, 0, packetLengthBytes.Length);
+                Buffer.BlockCopy(seqBytes, 0, rv, packetLengthBytes.Length, seqBytes.Length);
+                Buffer.BlockCopy(body, 0, rv, packetLengthBytes.Length + seqBytes.Length, body.Length);
+                var crc32 = new Crc32();
+                crc32.SlurpBlock(rv, 0, rv.Length);
+                var validChecksum = crc32.Crc32Result;
 
-            if (checksum != validChecksum)
+                if (checksum != validChecksum)
+                {
+                    throw new InvalidOperationException("invalid checksum! skip");
+                }
+
+                return body;
+            }
+            catch
             {
-                throw new InvalidOperationException("invalid checksum! skip");
+                var buffer = new byte[4];
+                while (stream.CanRead && stream.DataAvailable)

```

Comment:
@vik-borisov,
I wrapped it into a try catch and drained the tcp stream to get back to a good state. This is an ugly implementation and skips the malformed message without a retry. Maybe you have a better idea what can be done here